wdl-lint/src/rules/comment_whitespace.rs

```diff
@@ -0,0 +1,251 @@
+             indentation."
+        ))
+}
+
+/// Detects improperly spaced comments.
+#[derive(Default, Debug, Clone, Copy)]
+pub struct CommentWhitespaceRule;
+
+impl Rule for CommentWhitespaceRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that WDL comments have the proper spacing."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Comments on the same line as code should have 2 spaces before the # and one space before \
+         the comment text. Comments on their own line should match the indentation level around \
+         them and have one space between the # and the comment text. Keep in mind that even \
+         comments must be kept below the 90 character width limit."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Spacing])
+    }
+}
+
+impl Visitor for CommentWhitespaceRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn comment(&mut self, state: &mut Self::State, comment: &Comment) {
+        if is_inline_comment(comment) {
+            // check preceding whitespace for two spaces
+            if let Some(prior) = comment.syntax().prev_sibling_or_token() {
+                if prior.kind() != SyntaxKind::Whitespace || prior.to_string() != "  " {
+                    // Report a diagnostic if there are not two spaces before the comment delimiter
+                    state.add(inline_preceding_whitespace(comment.span()))
+                }
+            }
+        } else {
+            // Not an in-line comment, so check indentation level
+            let ancestors = comment
+                .syntax()
+                .parent_ancestors()
+                .filter(|a| {
+                    if let Some(prior) = a.prev_sibling_or_token() {
+                        if prior.kind() == SyntaxKind::Whitespace
+                            && prior.to_string().contains('\n')
+                        {
+                            return true;
+                        }
+                    }
+                    if let Some(next) = a.first_child_or_token() {
+                        if next.kind() == SyntaxKind::OpenParen {
+                            if let Some(next_next) = next.next_sibling_or_token() {
+                                if next_next.kind() == SyntaxKind::Whitespace
+                                    && next_next.to_string().contains('\n')
+                                {
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                    false
+                })
```

Comment:
Can you please refactor this chunk into its own function? `filter_parent_ancestors(a: SyntaxNode) -> bool` I think would be the signature. Asking so that it can be unit tested. I have a hunch that L131 needs to include `SyntaxKind::OpenBracket` and `OpenBrace` as well as `OpenParen` but I'm not totally sure.

Reply:
I pulled that logic out and put in a number of test cases. I updated the logic as well. I'm fairly certain that the special case only applies to `ParenthesizedExpressionNode`s because those are the only ones that can be part of a arithmetic expression, where a `\n` placed before the operator token appears inside the `ExpressionNode` itself. That seems to be the tricky case.

Reply:
What if an expression contains an array, map, or object? All those are possible, and wouldn't they contribute a level to indentation?

Reply:
None of those are permissible on the RHS of an expression (other than `=`, `==`, and `!=`) right?

Reply:
@a-frantz  - I added test cases with array and map/object expressions. I think everything is working as expected.